doc/sphinx/proof-engine/proof-handling.rst

```diff
@@ -539,6 +539,112 @@ Requesting information
+
+**How to enable diffs**
+
+.. opt:: Set Diffs %( "on" %| "off" %| "removed" %)
+.. This ref doesn't work: :opt:`Set Diffs %( "on" %| "off" %| "removed" %)`
+
+The “on” option highlights added tokens in green, while the “removed” option
+additionally reprints items with removed tokens in red.  Unchanged tokens in
+modified items are shown with pale green or red.  (Colors are user-configurable.)
+
+For coqtop, showing diffs can be enabled when starting coqtop with the
+``-diffs on|off|removed`` command-line option or with the ``Set Diffs``
+command within Coq.  You will need to provide the ``-color on|auto`` command-line option when
+you start coqtop in either case.
+
+Colors for coqtop can be configured by setting the ``COQ_COLORS`` environment
+variable, which has the same format as the GNU ``LS_COLORS`` variable.  The string
+uses ANSI escape codes to represent attributes.  For example:
+
+        ``export COQ_COLORS=”diff.added=4;48;2;0;0;240:diff.added.bg=44”``
+
+sets the highlights for added text to underlined (the 4) with a background RGB
+color (0, 0, 240) and unmodified text in a modified item to a blue background.
+The other tags used by diffs are ``diff.removed`` and ``diff.removed.bg``.  Note
+that if you specify ``COQ_COLORS``, the predefined attributes are ignored.
+
+
+In CoqIDE, diffs should be enabled from the View menu.  Don’t use the ``Set Diffs``
+command in CoqIDE.  You can change the background colors shown for diffs from the
+Edit | Preferences | Tags panel by changing the settings for the ``diff.added``,
+``diff.added.bg``, ``diff.removed`` and ``diff.removed.bg`` tags.  This panel also
+allows let you control other attributes of the highlights, such as the foreground
+color, bold, italic, underline and strikeout.
+
+Note: As of this writing (August 2018), Proof General will need minor changes
+to be able to show diffs correctly.  We hope it will support this feature soon.
+
+**How diffs are calculated**
+
+Diffs are calculated as follows:
+
+1. Select the old proof state to compare to, which is the proof state before
+   the last tactic that changed the proof.  Changes that only affect the view
+   of the proof, such as ``all: swap 1 2``, are ignored.
+
+2. For each goal in the new proof state, determine what old goal to compare
+   it to--the one it is derived from or is the same as.  Match the hypotheses by
+   name (order is ignored), handling compacted items specially.
+
+3. For each hypothesis and conclusion (the “items”) in each goal, pass
+   them as strings to the lexer to break them into tokens.  Then apply the standard
+   Myers diff algorithm to the tokens and add appropriate highlighting.
+
+Notes:
+
+* Aside from the highlights, the "on" option output should be identical
+  to the undiffed output.
+* Deleted goals (i.e. completed in the last proof step) will not be shown
+  even with the "removed" setting.
+* If you get a ``Diff_Failure`` message, that may be a bug.  While unlikely,
```

Comment:
I'm not sure what you are saying here. If it is an anomaly, then it *is* most definitely a bug and should be reported as such. I'm not sure about the worthiness of mentioning the `Back` command since nobody uses this AFAICT.

Reply:
I didn't have time to modify the lexer so that, for diffs, it accepts all strings.  That's next on my list.  @ejgallego and I thought it was pretty unlikely that something printed in the proof would not be lexable.  I do want to fix that, though, so the underlying diff code can later be used with any input strings, such as for error messages for type errors (diffing the expected and actual) or perhaps Gallina code.  Or other things.

I'll remove this.

Reply:
Oh, are diffs based on string comparisons?

Reply:
Yes, diff starts with strings, which it passes through the lexer to get a list of strings/tokens that the diff treats as opaque values--["abc"] compared to ["abcd"] gives one deletion ("abc") and one insertion ("abcd").  Whitespace differences are ignored since they are not tokens.